persistence/sql/persister_grant_jwk.go

```diff
@@ -0,0 +1,196 @@
+package sql
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/gobuffalo/pop/v5"
+	"gopkg.in/square/go-jose.v2"
+
+	"github.com/ory/hydra/oauth2/trust"
+	"github.com/ory/x/errorsx"
+
+	"github.com/ory/x/sqlcon"
+)
+
+var _ trust.GrantManager = &Persister{}
+
+const scopeSeparator = " "
+
+func (p *Persister) CreateGrant(ctx context.Context, g trust.Grant, publicKey jose.JSONWebKey) error {
+	// add key, if it doesn't exist
+	if _, err := p.GetKey(ctx, g.PublicKey.Set, g.PublicKey.KeyID); err != nil {
+		if errorsx.Cause(err) != sqlcon.ErrNoRows {
+			return err
+		}
+
+		if err = p.AddKey(ctx, g.PublicKey.Set, &publicKey); err != nil {
+			return err
+		}
+	}
+
+	data := p.sqlDataFromJWTGrant(g)
+
+	return sqlcon.HandleError(p.Connection(ctx).Create(&data))
```

Comment:
This should be wrapped in one transaction, otherwise `Create` might fail, but we already added the key to the store.

Reply:
Refactored in https://github.com/TinkoffCreditSystems/hydra/pull/3